File: src/syntax/bibtex/parser.ts
import { TokenBuffer } from '../token';
import {
  BibtexBracedContentSyntax,
  BibtexCommandSyntax,
  BibtexCommentSyntax,
  BibtexConcatSyntax,
  BibtexContentSyntax,
  BibtexDocumentItemSyntax,
  BibtexDocumentSyntax,
  BibtexEntrySyntax,
  BibtexFieldSyntax,
  BibtexPreambleSyntax,
  BibtexQuotedContentSyntax,
  BibtexStringSyntax,
  BibtexWordSyntax,
} from './ast';
import { BibtexLexer, BibtexToken, BibtexTokenKind } from './lexer';

class BibtexParser {
  constructor(private readonly tokens: TokenBuffer<BibtexToken>) {}

  public document(): BibtexDocumentSyntax {
    const children: BibtexDocumentItemSyntax[] = [];
    while (this.tokens.available) {
      switch (this.tokens.peek()!.kind) {
        case BibtexTokenKind.PreambleType:
          children.push(this.preamble());
          break;
        case BibtexTokenKind.StringType:
          children.push(this.string());
          break;
        case BibtexTokenKind.EntryType:
          children.push(this.entry());
          break;
        default:
          const token = this.tokens.next();
          children.push(new BibtexCommentSyntax(token));
          break;
      }
    }
    return new BibtexDocumentSyntax(children);
  }

  private preamble(): BibtexPreambleSyntax {
    const type = this.tokens.next();
    const left = this.expect(
      BibtexTokenKind.BeginBrace,
      BibtexTokenKind.BeginParen,
    );
    if (left === undefined) {
      return new BibtexPreambleSyntax(type, undefined, undefined, undefined);
    }

    if (!this.canMatchContent()) {
      return new BibtexPreambleSyntax(type, left, undefined, undefined);
    }
    const content = this.content();

    const right = this.expect(
      BibtexTokenKind.EndBrace,
      BibtexTokenKind.EndParen,
    );
    return new BibtexPreambleSyntax(type, left, content, right);
  }

  private string(): BibtexStringSyntax {
    const type = this.tokens.next();

    const left = this.expect(
      BibtexTokenKind.BeginBrace,
      BibtexTokenKind.BeginParen,
    );
    if (left === undefined) {
      return new BibtexStringSyntax(
        type,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      );
    }

    const name = this.expect(BibtexTokenKind.Word);
    if (name === undefined) {
      return new BibtexStringSyntax(
        type,
        left,
        undefined,
        undefined,
        undefined,
        undefined,
      );
    }

    const assign = this.expect(BibtexTokenKind.Assign);
    if (assign === undefined) {
      return new BibtexStringSyntax(
        type,
        left,
        name,
        undefined,
        undefined,
        undefined,
      );
    }

    if (!this.canMatchContent()) {
      return new BibtexStringSyntax(
        type,
        left,
        name,
        assign,
        undefined,
        undefined,
      );
    }
    const value = this.content();

    const right = this.expect(
      BibtexTokenKind.EndBrace,
      BibtexTokenKind.EndParen,
    );
    return new BibtexStringSyntax(type, left, name, assign, value, right);
  }

  private entry(): BibtexEntrySyntax {
    const type = this.tokens.next();

    const left = this.expect(
      BibtexTokenKind.BeginBrace,
      BibtexTokenKind.BeginParen,
    );
    if (left === undefined) {
      return new BibtexEntrySyntax(
        type,
        undefined,
        undefined,
        undefined,
        [],
        undefined,
      );
    }

    const name = this.expect(BibtexTokenKind.Word);
    if (name === undefined) {
      return new BibtexEntrySyntax(
        type,
        left,
        undefined,
        undefined,
        [],
        undefined,
      );
    }

    const comma = this.expect(BibtexTokenKind.Comma);
    if (comma === undefined) {
      return new BibtexEntrySyntax(type, left, name, undefined, [], undefined);
    }

    const fields: BibtexFieldSyntax[] = [];
    while (this.nextOfKind(BibtexTokenKind.Word)) {
      fields.push(this.field());
    }

    const right = this.expect(
      BibtexTokenKind.EndBrace,
      BibtexTokenKind.EndParen,
    );

    return new BibtexEntrySyntax(type, left, name, comma, fields, right);
  }

  private field(): BibtexFieldSyntax {
    const name = this.tokens.next();
    const assign = this.expect(BibtexTokenKind.Assign);
    if (assign === undefined) {
      return new BibtexFieldSyntax(name, undefined, undefined, undefined);
    }

    if (!this.canMatchContent()) {
      return new BibtexFieldSyntax(name, assign, undefined, undefined);
    }
    const content = this.content();

    const comma = this.expect(BibtexTokenKind.Comma);
    return new BibtexFieldSyntax(name, assign, content, comma);
  }

  private content(): BibtexContentSyntax {
    let left: BibtexContentSyntax;
    const token = this.tokens.next();
    switch (token.kind) {
      case BibtexTokenKind.PreambleType:
      case BibtexTokenKind.StringType:
      case BibtexTokenKind.EntryType:
      case BibtexTokenKind.Word:
      case BibtexTokenKind.Assign:
      case BibtexTokenKind.Comma:
      case BibtexTokenKind.BeginParen:
      case BibtexTokenKind.EndParen:
        left = new BibtexWordSyntax(token);
        break;
      case BibtexTokenKind.Command:
        left = new BibtexCommandSyntax(token);
        break;
      case BibtexTokenKind.Quote: {
        const children: BibtexContentSyntax[] = [];
        while (this.canMatchContent()) {
          const kind = this.tokens.peek()!.kind;
          if (kind === BibtexTokenKind.Quote) {
            break;
          }
          children.push(this.content());
        }
        const right = this.expect(BibtexTokenKind.Quote);
        left = new BibtexQuotedContentSyntax(token, children, right);
        break;
      }
      case BibtexTokenKind.BeginBrace: {
        const children: BibtexContentSyntax[] = [];
        while (this.canMatchContent()) {
          children.push(this.content());
        }
        const right = this.expect(BibtexTokenKind.EndBrace);
        left = new BibtexBracedContentSyntax(token, children, right);
        break;
      }
      default:
        throw Error('Unexpected token type: ' + token.kind);
    }
    const operator = this.expect(BibtexTokenKind.Concat);
    if (operator === undefined) {
      return left;
    } else {
      const right = this.canMatchContent() ? this.content() : undefined;
      return new BibtexConcatSyntax(left, operator, right);
    }
  }

  private canMatchContent(): boolean {
    const token = this.tokens.peek();
    if (token === undefined) {
      return false;
    }

    switch (token.kind) {
      case BibtexTokenKind.PreambleType:
      case BibtexTokenKind.StringType:
      case BibtexTokenKind.EntryType:
      case BibtexTokenKind.Word:
      case BibtexTokenKind.Command:
      case BibtexTokenKind.Assign:
      case BibtexTokenKind.Comma:
        return true;
      case BibtexTokenKind.Concat:
        return false;
      case BibtexTokenKind.Quote:
        return true;
      case BibtexTokenKind.BeginBrace:
        return true;
      case BibtexTokenKind.EndBrace:
        return false;
      case BibtexTokenKind.BeginParen:
      case BibtexTokenKind.EndParen:
        return true;
    }
  }

  private expect(...kinds: BibtexTokenKind[]): BibtexToken | undefined {
    const token = this.tokens.peek();
    return token !== undefined && kinds.some(x => x === token.kind)
      ? this.tokens.next()
      : undefined;
  }

  private nextOfKind(kind: BibtexTokenKind): boolean {
    const token = this.tokens.peek();
    return token !== undefined && token.kind === kind;
  }
}

export function parse(text: string): BibtexDocumentSyntax {
  const lexer = new BibtexLexer(text);
  const tokens = new TokenBuffer(lexer);
  const parser = new BibtexParser(tokens);
  return parser.document();
}

File: crates/texlab_syntax/src/bibtex/parser.rs
use super::ast::*;
use crate::text::SyntaxNode;
use petgraph::graph::{Graph, NodeIndex};
use std::iter::Peekable;
use texlab_protocol::{Range, RangeExt};

pub struct Parser<I: Iterator<Item = Token>> {
    graph: Graph<Node, ()>,
    tokens: Peekable<I>,
}

impl<I: Iterator<Item = Token>> Parser<I> {
    pub fn new(tokens: I) -> Self {
        Self {
            tokens: tokens.peekable(),
            graph: Graph::new(),
        }
    }

    pub fn parse(mut self) -> Tree {
        let mut children = Vec::new();

        while let Some(ref token) = self.tokens.peek() {
            match token.kind {
                TokenKind::PreambleKind => children.push(self.preamble()),
                TokenKind::StringKind => children.push(self.string()),
                TokenKind::EntryKind => children.push(self.entry()),
                _ => children.push(self.comment()),
            }
        }

        let range = if children.is_empty() {
            Range::new_simple(0, 0, 0, 0)
        } else {
            let start = self.graph[children[0]].start();
            let end = self.graph[children[children.len() - 1]].end();
            Range::new(start, end)
        };

        let root = self.graph.add_node(Node::Root(Root { range }));
        self.connect(root, &children);
        Tree {
            graph: self.graph,
            root,
        }
    }

    fn preamble(&mut self) -> NodeIndex {
        let ty = self.tokens.next().unwrap();

        let left = self.expect2(TokenKind::BeginBrace, TokenKind::BeginParen);
        if left.is_none() {
            return self.graph.add_node(Node::Preamble(Box::new(Preamble {
                range: ty.range(),
                ty,
                left: None,
                right: None,
            })));
        }

        if !self.can_match_content() {
            return self.graph.add_node(Node::Preamble(Box::new(Preamble {
                range: Range::new(ty.start(), left.as_ref().unwrap().end()),
                ty,
                left,
                right: None,
            })));
        }

        let content = self.content();

        let right = self.expect2(TokenKind::EndBrace, TokenKind::EndParen);
        let end = right
            .as_ref()
            .map(Token::end)
            .unwrap_or_else(|| self.graph[content].end());

        let parent = self.graph.add_node(Node::Preamble(Box::new(Preamble {
            range: Range::new(ty.start(), end),
            ty,
            left,
            right,
        })));
        self.graph.add_edge(parent, content, ());
        parent
    }

    fn string(&mut self) -> NodeIndex {
        let ty = self.tokens.next().unwrap();

        let left = self.expect2(TokenKind::BeginBrace, TokenKind::BeginParen);
        if left.is_none() {
            return self.graph.add_node(Node::String(Box::new(String {
                range: ty.range(),
                ty,
                left: None,
                name: None,
                assign: None,
                right: None,
            })));
        }

        let name = self.expect1(TokenKind::Word);
        if name.is_none() {
            return self.graph.add_node(Node::String(Box::new(String {
                range: Range::new(ty.start(), left.as_ref().unwrap().end()),
                ty,
                left,
                name: None,
                assign: None,
                right: None,
            })));
        }

        let assign = self.expect1(TokenKind::Assign);
        if assign.is_none() {
            return self.graph.add_node(Node::String(Box::new(String {
                range: Range::new(ty.start(), name.as_ref().unwrap().end()),
                ty,
                left,
                name,
                assign: None,
                right: None,
            })));
        }

        if !self.can_match_content() {
            return self.graph.add_node(Node::String(Box::new(String {
                range: Range::new(ty.start(), assign.as_ref().unwrap().end()),
                ty,
                left,
                name,
                assign,
                right: None,
            })));
        }
        let value = self.content();

        let right = self.expect2(TokenKind::EndBrace, TokenKind::EndParen);
        let end = right
            .as_ref()
            .map(Token::end)
            .unwrap_or_else(|| self.graph[value].end());

        let parent = self.graph.add_node(Node::String(Box::new(String {
            range: Range::new(ty.start(), end),
            ty,
            left,
            name,
            assign,
            right,
        })));
        self.graph.add_edge(parent, value, ());
        parent
    }

    fn entry(&mut self) -> NodeIndex {
        let ty = self.tokens.next().unwrap();

        let left = self.expect2(TokenKind::BeginBrace, TokenKind::BeginParen);
        if left.is_none() {
            return self.graph.add_node(Node::Entry(Box::new(Entry {
                range: ty.range(),
                ty,
                left: None,
                key: None,
                comma: None,
                right: None,
            })));
        }

        let key = self.expect1(TokenKind::Word);
        if key.is_none() {
            return self.graph.add_node(Node::Entry(Box::new(Entry {
                range: Range::new(ty.start(), left.as_ref().unwrap().end()),
                ty,
                left,
                key: None,
                comma: None,
                right: None,
            })));
        }

        let comma = self.expect1(TokenKind::Comma);
        if comma.is_none() {
            return self.graph.add_node(Node::Entry(Box::new(Entry {
                range: Range::new(ty.start(), key.as_ref().unwrap().end()),
                ty,
                left,
                key,
                comma: None,
                right: None,
            })));
        }

        let mut fields = Vec::new();
        while self.next_of_kind(TokenKind::Word) {
            fields.push(self.field());
        }

        let right = self.expect2(TokenKind::EndBrace, TokenKind::EndParen);

        let end = right
            .as_ref()
            .map(Token::end)
            .or_else(|| fields.last().map(|field| self.graph[*field].end()))
            .unwrap_or_else(|| comma.as_ref().unwrap().end());
        let parent = self.graph.add_node(Node::Entry(Box::new(Entry {
            range: Range::new(ty.start(), end),
            ty,
            left,
            key,
            comma,
            right,
        })));
        self.connect(parent, &fields);
        parent
    }

    fn comment(&mut self) -> NodeIndex {
        let token = self.tokens.next().unwrap();
        self.graph.add_node(Node::Comment(Comment { token }))
    }

    fn field(&mut self) -> NodeIndex {
        let name = self.tokens.next().unwrap();

        let assign = self.expect1(TokenKind::Assign);
        if assign.is_none() {
            return self.graph.add_node(Node::Field(Box::new(Field {
                range: name.range(),
                name,
                assign: None,
                comma: None,
            })));
        }

        if !self.can_match_content() {
            return self.graph.add_node(Node::Field(Box::new(Field {
                range: Range::new(name.start(), assign.as_ref().unwrap().end()),
                name,
                assign,
                comma: None,
            })));
        }

        let content = self.content();

        let comma = self.expect1(TokenKind::Comma);

        let end = comma
            .as_ref()
            .map(Token::end)
            .unwrap_or_else(|| self.graph[content].end());
        let parent = self.graph.add_node(Node::Field(Box::new(Field {
            range: Range::new(name.start(), end),
            name,
            assign,
            comma,
        })));
        self.graph.add_edge(parent, content, ());
        parent
    }

    fn content(&mut self) -> NodeIndex {
        let token = self.tokens.next().unwrap();
        let left = match token.kind {
            TokenKind::PreambleKind
            | TokenKind::StringKind
            | TokenKind::EntryKind
            | TokenKind::Word
            | TokenKind::Assign
            | TokenKind::Comma
            | TokenKind::BeginParen
            | TokenKind::EndParen => self.graph.add_node(Node::Word(Word { token })),
            TokenKind::Command => self.graph.add_node(Node::Command(Command { token })),
            TokenKind::Quote => {
                let mut children = Vec::new();
                while self.can_match_content() {
                    if self.next_of_kind(TokenKind::Quote) {
                        break;
                    }
                    children.push(self.content());
                }
                let right = self.expect1(TokenKind::Quote);

                let end = right
                    .as_ref()
                    .map(Token::end)
                    .or_else(|| children.last().map(|child| self.graph[*child].end()))
                    .unwrap_or_else(|| token.end());
                let parent = self.graph.add_node(Node::QuotedContent(QuotedContent {
                    range: Range::new(token.start(), end),
                    left: token,
                    right,
                }));
                self.connect(parent, &children);
                parent
            }
            TokenKind::BeginBrace => {
                let mut children = Vec::new();
                while self.can_match_content() {
                    children.push(self.content());
                }
                let right = self.expect1(TokenKind::EndBrace);

                let end = right
                    .as_ref()
                    .map(Token::end)
                    .or_else(|| children.last().map(|child| self.graph[*child].end()))
                    .unwrap_or_else(|| token.end());
                let parent = self.graph.add_node(Node::BracedContent(BracedContent {
                    range: Range::new(token.start(), end),
                    left: token,
                    right,
                }));
                self.connect(parent, &children);
                parent
            }
            _ => unreachable!(),
        };

        match self.expect1(TokenKind::Concat) {
            Some(operator) => {
                if self.can_match_content() {
                    let right = self.content();
                    let parent = self.graph.add_node(Node::Concat(Concat {
                        range: Range::new(self.graph[left].start(), self.graph[right].end()),
                        operator,
                    }));
                    self.graph.add_edge(parent, left, ());
                    self.graph.add_edge(parent, right, ());
                    parent
                } else {
                    let parent = self.graph.add_node(Node::Concat(Concat {
                        range: Range::new(self.graph[left].start(), operator.end()),
                        operator,
                    }));
                    self.graph.add_edge(parent, left, ());
                    parent
                }
            }
            None => left,
        }
    }

    fn connect(&mut self, parent: NodeIndex, children: &[NodeIndex]) {
        for child in children {
            self.graph.add_edge(parent, *child, ());
        }
    }

    fn can_match_content(&mut self) -> bool {
        if let Some(ref token) = self.tokens.peek() {
            match token.kind {
                TokenKind::PreambleKind
                | TokenKind::StringKind
                | TokenKind::EntryKind
                | TokenKind::Word
                | TokenKind::Command
                | TokenKind::Assign
                | TokenKind::Comma
                | TokenKind::Quote
                | TokenKind::BeginBrace
                | TokenKind::BeginParen
                | TokenKind::EndParen => true,
                TokenKind::Concat | TokenKind::EndBrace => false,
            }
        } else {
            false
        }
    }

    fn expect1(&mut self, kind: TokenKind) -> Option<Token> {
        if let Some(ref token) = self.tokens.peek() {
            if token.kind == kind {
                return self.tokens.next();
            }
        }
        None
    }

    fn expect2(&mut self, kind1: TokenKind, kind2: TokenKind) -> Option<Token> {
        if let Some(ref token) = self.tokens.peek() {
            if token.kind == kind1 || token.kind == kind2 {
                return self.tokens.next();
            }
        }
        None
    }

    fn next_of_kind(&mut self, kind: TokenKind) -> bool {
        if let Some(token) = self.tokens.peek() {
            token.kind == kind
        } else {
            false
        }
    }
}
